[Bug] Fix the broken CI

## problem 1: android

https://github.com/fzyzcjy/flutter_rust_bridge/runs/5137000331?check_suite_focus=true

```
[+31113 ms] > Configure project :integration_test
[        ] Warning: The plugin integration_test requires Android SDK version 31.
[   +1 ms] One or more plugins require a higher Android SDK version.
[   +2 ms] Fix this issue by adding the following to /Users/runner/work/flutter_rust_bridge/flutter_rust_bridge/frb_example/with_flutter/android/app/build.gradle:
[        ] android {
[        ]   compileSdkVersion 31
[        ]   ...
[        ] }

...

[ +101 ms] > Task :app:mergeDebugResources
[        ] Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
[        ] Use '--warning-mode all' to show the individual deprecation warnings.
[        ] See https://docs.gradle.org/6.7/userguide/command_line_interface.html#sec:command_line_warnings
[        ] 15 actionable tasks: 14 executed, 1 up-to-date
[        ] FAILURE: Build failed with an exception.
[        ] * What went wrong:
[        ] Execution failed for task ':app:checkDebugAarMetadata'.
[        ] > Multiple task action failures occurred:
[        ]    > A failure occurred while executing com.android.build.gradle.internal.tasks.CheckAarMetadataWorkAction
[        ]       > The minCompileSdk (31) specified in a
[        ]         dependency's AAR metadata (META-INF/com/android/build/gradle/aar-metadata.properties)
[        ]         is greater than this module's compileSdkVersion (android-30).
[        ]         Dependency: androidx.window:window-java:1.0.0-beta04.
[        ]         AAR metadata file: /Users/runner/.gradle/caches/transforms-2/files-2.1/625039eaad011f884ddd84f857a44b7f/jetified-window-java-1.0.0-beta04/META-INF/com/android/build/gradle/aar-metadata.properties.
[        ]    > A failure occurred while executing com.android.build.gradle.internal.tasks.CheckAarMetadataWorkAction
[        ]       > The minCompileSdk (31) specified in a
[        ]         dependency's AAR metadata (META-INF/com/android/build/gradle/aar-metadata.properties)
[        ]         is greater than this module's compileSdkVersion (android-30).
[        ]         Dependency: androidx.window:window:1.0.0-beta04.
[        ]         AAR metadata file: /Users/runner/.gradle/caches/transforms-2/files-2.1/a78fdf90e4c1f8464b19895cfb365f3f/jetified-window-1.0.0-beta04/META-INF/com/android/build/gradle/aar-metadata.properties.
```

later see:

```
           ┌─ Flutter Fix ────────────────────────────────────────────────────────────────────────────────────┐
           │ [!] Your project requires a newer version of the Kotlin Gradle plugin.                           │
           │ Find the latest version on https://kotlinlang.org/docs/gradle.html#plugin-and-versions, then     │
           │ update                                                                                           │
           │ /Users/runner/work/flutter_rust_bridge/flutter_rust_bridge/frb_example/with_flutter/android/buil │
           │ d.gradle:                                                                                        │
           │ ext.kotlin_version = '<latest-version>'                                                          │
           └──────────────────────────────────────────────────────────────────────────────────────────────────┘

```

## problem 2: windows

https://github.com/fzyzcjy/flutter_rust_bridge/runs/5137000113?check_suite_focus=true

```
[2022-02-10T07:07:26Z WARN  flutter_rust_bridge_codegen::commands] command="cmd" "/C" "rustfmt \"D:\\a\\flutter_rust_bridge\\flutter_rust_bridge\\frb_codegen\\..\\\\frb_example\\\\pure_dart\\\\rust\\\\src\\bridge_generated.rs\"" stdout= stderr=Error: file `"D:\a\flutter_rust_bridge\flutter_rust_bridge\frb_codegen\..\\frb_example\\pure_dart\\rust\\src\bridge_generated.rs"` does not exist
```

--- a/frb_codegen/src/commands.rs
+++ b/frb_codegen/src/commands.rs
@@ -127,7 +127,7 @@ fn execute_command(bin: &str, args: &[&str], current_dir: Option<&str>) -> Outpu
             String::from_utf8_lossy(&result.stderr)
         );
     }
-    return result;
+    result
 }
 
 fn cbindgen(rust_crate_dir: &str, c_output_path: &str, c_struct_names: Vec<String>) {
@@ -220,7 +220,7 @@ fn ffigen(
         )
         .unwrap();
         for path in llvm_path {
-            write!(&mut config, "           - '{}'\n", path).unwrap();
+            writeln!(&mut config, "           - '{}'", path).unwrap();
         }
     }
 

--- a/frb_codegen/src/commands.rs
+++ b/frb_codegen/src/commands.rs
@@ -16,9 +16,9 @@ enum Failures {
 }
 
 #[must_use]
-fn call_shell(cmd: &'static str) -> Output {
+fn call_shell(cmd: &str) -> Output {
     #[cfg(windows)]
-    return execute_command("cmd", &["/C", cmd], None);
+    return execute_command("powershell", &["-noprofile", "-c", cmd], None);
 
     #[cfg(not(windows))]
     execute_command("sh", &["-c", cmd], None)
@@ -242,18 +242,10 @@ fn ffigen(
     debug!("ffigen config_file: {:?}", config_file);
 
     // NOTE please install ffigen globally first: `dart pub global activate ffigen`
-    let res = execute_command(
-        "dart",
-        &[
-            "pub",
-            "global",
-            "run",
-            "ffigen",
-            "--config",
-            &config_file.path().to_string_lossy(),
-        ],
-        None,
-    );
+    let res = call_shell(&format!(
+        "dart pub global run ffigen --config \"{}\"",
+        config_file.path().to_string_lossy()
+    ));
     if !res.status.success() {
         let err = String::from_utf8_lossy(&res.stderr);
         if err.contains("Couldn't find dynamic library in default locations.") {
@@ -284,11 +276,10 @@ pub fn format_dart(path: &str, line_length: i32) {
         "execute format_dart path={} line_length={}",
         path, line_length
     );
-    let res = execute_command(
-        "dart",
-        &["format", path, "--line-length", &line_length.to_string()],
-        None,
-    );
+    let res = call_shell(&format!(
+        "dart format {} --line-length {}",
+        path, line_length
+    ));
     if !res.status.success() {
         error!(
             "dart format failed: {}",

--- a/frb_codegen/src/commands.rs
+++ b/frb_codegen/src/commands.rs
@@ -55,7 +55,7 @@ pub fn check_shell_executable(cmd: &str) {
     // TODO: Implement check_shell_executable on Windows
     let res = execute_command(
         "powershell",
-        &["-Command", &format!("& { Get-Command -Name {} }", cmd)],
+        &["-Command", &format!("& {{ Get-Command -Name {} }}", cmd)],
         None,
     );
     todo!("{:#?}", res);

--- a/frb_codegen/src/commands.rs
+++ b/frb_codegen/src/commands.rs
@@ -52,8 +52,13 @@ Note: This command might be available via cargo, in which case it can be install
 
 #[cfg(windows)]
 pub fn check_shell_executable(cmd: &str) {
-    warn!("check_shell_executable not implemented on Windows");
     // TODO: Implement check_shell_executable on Windows
+    let res = execute_command(
+        "powershell",
+        &["-Command", &format!("& { Get-Command -Name {} }", cmd)],
+        None,
+    );
+    todo!("{:#?}", res);
 }
 
 pub fn bindgen_rust_to_dart(

--- a/frb_codegen/src/commands.rs
+++ b/frb_codegen/src/commands.rs
@@ -26,7 +26,6 @@ ffigen is not available, please run \"dart pub global activate ffigen\" first."
         std::process::exit(Failures::MissingExe as _);
     }
 
-    check_shell_executable("flutter_rust_bridge_codegen");
     check_shell_executable("cbindgen");
 }
 

--- a/frb_codegen/src/commands.rs
+++ b/frb_codegen/src/commands.rs
@@ -15,8 +15,17 @@ enum Failures {
     MissingExe,
 }
 
+#[must_use]
+fn call_shell(cmd: &'static str) -> Output {
+    #[cfg(windows)]
+    return execute_command("cmd", &["/C", cmd], None);
+
+    #[cfg(not(windows))]
+    execute_command("sh", &["-c", cmd], None)
+}
+
 pub fn ensure_tools_available() {
-    let output = execute_command("dart", &["pub", "global", "list"], None);
+    let output = call_shell("dart pub global list");
     let output = String::from_utf8_lossy(&output.stdout);
     if !output.contains("ffigen") {
         error!(
@@ -29,13 +38,9 @@ ffigen is not available, please run \"dart pub global activate ffigen\" first."
     check_shell_executable("cbindgen");
 }
 
-pub fn check_shell_executable(cmd: &str) {
+pub fn check_shell_executable(cmd: &'static str) {
     #[cfg(windows)]
-    let res = execute_command(
-        "powershell",
-        &["-c", &format!("Get-Command -Name {}", cmd)],
-        None,
-    );
+    let res = execute_command("where", &[cmd], None);
     #[cfg(not(windows))]
     let res = execute_command(
         "sh",
@@ -48,12 +53,12 @@ pub fn check_shell_executable(cmd: &str) {
 {cmd} is not a command, or not executable.
 Note: This command might be available via cargo, in which case it can be installed with:
 
-    cargo install {cmd}
-",
+    cargo install {cmd}",
             cmd = cmd
         );
         std::process::exit(Failures::MissingExe as _);
     }
+    debug!("{}", String::from_utf8_lossy(&res.stdout));
 }
 
 pub fn bindgen_rust_to_dart(
@@ -78,7 +83,6 @@ pub fn bindgen_rust_to_dart(
 #[must_use = "Error path must be handled."]
 fn execute_command(bin: &str, args: &[&str], current_dir: Option<&str>) -> Output {
     let mut cmd = Command::new(bin);
-
     cmd.args(args);
 
     if let Some(current_dir) = current_dir {
@@ -92,11 +96,10 @@ fn execute_command(bin: &str, args: &[&str], current_dir: Option<&str>) -> Outpu
 
     let result = cmd
         .output()
-        .unwrap_or_else(|err| panic!("\"{} {}\" failed: {}", bin, args.join(" "), err));
+        .unwrap_or_else(|err| panic!("\"{}\" \"{}\" failed: {}", bin, args.join(" "), err));
 
+    let stdout = String::from_utf8_lossy(&result.stdout);
     if result.status.success() {
-        let stdout = String::from_utf8_lossy(&result.stdout);
-
         debug!(
             "command={:?} stdout={} stderr={}",
             cmd,
@@ -122,7 +125,7 @@ fn execute_command(bin: &str, args: &[&str], current_dir: Option<&str>) -> Outpu
         warn!(
             "command={:?} stdout={} stderr={}",
             cmd,
-            String::from_utf8_lossy(&result.stdout),
+            stdout,
             String::from_utf8_lossy(&result.stderr)
         );
     }
@@ -247,7 +250,7 @@ fn ffigen(
             "run",
             "ffigen",
             "--config",
-            config_file.path().to_str().unwrap(),
+            &config_file.path().to_string_lossy(),
         ],
         None,
     );
@@ -259,8 +262,7 @@ fn ffigen(
 ffigen could not find LLVM.
 Please supply --llvm-path to flutter_rust_bridge_codegen, e.g.:
 
-    flutter_rust_bridge_codegen .. --llvm-path <path_to_llvm>
-"
+    flutter_rust_bridge_codegen .. --llvm-path <path_to_llvm>"
             );
             std::process::exit(Failures::FfigenLlvm as _);
         }

--- a/frb_codegen/src/commands.rs
+++ b/frb_codegen/src/commands.rs
@@ -19,8 +19,10 @@ pub fn ensure_tools_available() {
     let output = execute_command("dart", &["pub", "global", "list"], None);
     let output = String::from_utf8_lossy(&output.stdout);
     if !output.contains("ffigen") {
-        error!("
-ffigen is not available, please run \"dart pub global activate ffigen\" first.");
+        error!(
+            "
+ffigen is not available, please run \"dart pub global activate ffigen\" first."
+        );
         std::process::exit(Failures::MissingExe as _);
     }
 
@@ -30,7 +32,11 @@ ffigen is not available, please run \"dart pub global activate ffigen\" first.")
 
 #[cfg(not(windows))]
 pub fn check_shell_executable(cmd: &str) {
-    let res = execute_command("sh", &["-c", &format!("test -x \"$(which {})\"", cmd)], None);
+    let res = execute_command(
+        "sh",
+        &["-c", &format!("test -x \"$(which {})\"", cmd)],
+        None,
+    );
     if !res.status.success() {
         error!(
             "

--- a/frb_codegen/src/commands.rs
+++ b/frb_codegen/src/commands.rs
@@ -29,8 +29,14 @@ ffigen is not available, please run \"dart pub global activate ffigen\" first."
     check_shell_executable("cbindgen");
 }
 
-#[cfg(not(windows))]
 pub fn check_shell_executable(cmd: &str) {
+    #[cfg(windows)]
+    let res = execute_command(
+        "powershell",
+        &["-c", &format!("Get-Command -Name {}", cmd)],
+        None,
+    );
+    #[cfg(not(windows))]
     let res = execute_command(
         "sh",
         &["-c", &format!("test -x \"$(which {})\"", cmd)],
@@ -48,17 +54,6 @@ Note: This command might be available via cargo, in which case it can be install
         );
         std::process::exit(Failures::MissingExe as _);
     }
-}
-
-#[cfg(windows)]
-pub fn check_shell_executable(cmd: &str) {
-    // TODO: Implement check_shell_executable on Windows
-    let res = execute_command(
-        "powershell",
-        &["-Command", &format!("& {{ Get-Command -Name {} }}", cmd)],
-        None,
-    );
-    todo!("{:#?}", res);
 }
 
 pub fn bindgen_rust_to_dart(

--- a/frb_codegen/src/commands.rs
+++ b/frb_codegen/src/commands.rs
@@ -1,8 +1,55 @@
-use std::io::Write;
+use std::fmt::Write;
 use std::path::Path;
 use std::process::Command;
+use std::process::Output;
 
-use log::{debug, warn};
+use log::{debug, error, warn};
+
+/// Known failures that occur from external commands.
+/// If an error occurs frequently enough, consider adding it here and use
+/// [std::process::exit] explicitly instead of panicking.
+enum Failures {
+    Rustfmt = 1,
+    Dartfmt,
+    FfigenLlvm,
+    MissingExe,
+}
+
+pub fn ensure_tools_available() {
+    let output = execute_command("dart", &["pub", "global", "list"], None);
+    let output = String::from_utf8_lossy(&output.stdout);
+    if !output.contains("ffigen") {
+        error!("
+ffigen is not available, please run \"dart pub global activate ffigen\" first.");
+        std::process::exit(Failures::MissingExe as _);
+    }
+
+    check_shell_executable("flutter_rust_bridge_codegen");
+    check_shell_executable("cbindgen");
+}
+
+#[cfg(not(windows))]
+pub fn check_shell_executable(cmd: &str) {
+    let res = execute_command("sh", &["-c", &format!("test -x \"$(which {})\"", cmd)], None);
+    if !res.status.success() {
+        error!(
+            "
+{cmd} is not a command, or not executable.
+Note: This command might be available via cargo, in which case it can be installed with:
+
+    cargo install {cmd}
+",
+            cmd = cmd
+        );
+        std::process::exit(Failures::MissingExe as _);
+    }
+}
+
+#[cfg(windows)]
+pub fn check_shell_executable(cmd: &str) {
+    warn!("check_shell_executable not implemented on Windows");
+    // TODO: Implement check_shell_executable on Windows
+}
 
 pub fn bindgen_rust_to_dart(
     rust_crate_dir: &str,
@@ -10,7 +57,7 @@ pub fn bindgen_rust_to_dart(
     dart_output_path: &str,
     dart_class_name: &str,
     c_struct_names: Vec<String>,
-    llvm_install_path: &str,
+    llvm_install_path: &[String],
     llvm_compiler_opts: &str,
 ) {
     cbindgen(rust_crate_dir, c_output_path, c_struct_names);
@@ -23,29 +70,24 @@ pub fn bindgen_rust_to_dart(
     );
 }
 
-fn execute_command(arg: &str, current_dir: Option<&str>) {
-    let mut cmd = if cfg!(target_os = "windows") {
-        let mut cmd = Command::new("cmd");
-        cmd.arg("/C");
-        cmd
-    } else {
-        let mut cmd = Command::new("sh");
-        cmd.arg("-c");
-        cmd
-    };
+#[must_use = "Error path must be handled."]
+fn execute_command(bin: &str, args: &[&str], current_dir: Option<&str>) -> Output {
+    let mut cmd = Command::new(bin);
 
-    cmd.arg(arg);
+    cmd.args(args);
 
     if let Some(current_dir) = current_dir {
         cmd.current_dir(current_dir);
     }
 
     debug!(
-        "execute command: arg={:?} current_dir={:?} cmd={:?}",
-        arg, current_dir, cmd
+        "execute command: bin={} args={:?} current_dir={:?} cmd={:?}",
+        bin, args, current_dir, cmd
     );
 
-    let result = cmd.output().unwrap();
+    let result = cmd
+        .output()
+        .unwrap_or_else(|err| panic!("\"{} {}\" failed: {}", bin, args.join(" "), err));
 
     if result.status.success() {
         let stdout = String::from_utf8_lossy(&result.stdout);
@@ -57,8 +99,8 @@ fn execute_command(arg: &str, current_dir: Option<&str>) {
             String::from_utf8_lossy(&result.stderr)
         );
         if stdout.contains("fatal error") {
-            warn!( "See keywords such as `error` in command output. Maybe there is a problem? command={:?} output={:?}", cmd, result);
-        } else if arg.contains("ffigen") && stdout.contains("[SEVERE]") {
+            warn!("See keywords such as `error` in command output. Maybe there is a problem? command={:?} output={:?}", cmd, result);
+        } else if args.contains(&"ffigen") && stdout.contains("[SEVERE]") {
             // HACK: If ffigen can't find a header file it will generate broken
             // bindings but still exit successfully. We can detect these broken
             // bindings by looking for a "[SEVERE]" log message.
@@ -67,8 +109,8 @@ fn execute_command(arg: &str, current_dir: Option<&str>) {
             // we don't want to error out completely.
 
             warn!(
-                "The `ffigen` command emitted a SEVERE error. Maybe there is a problem? command={:?} output={:?}",
-                cmd, result
+                "The `ffigen` command emitted a SEVERE error. Maybe there is a problem? command={:?} output=\n{}",
+                cmd, String::from_utf8_lossy(&result.stdout)
             );
         }
     } else {
@@ -78,8 +120,8 @@ fn execute_command(arg: &str, current_dir: Option<&str>) {
             String::from_utf8_lossy(&result.stdout),
             String::from_utf8_lossy(&result.stderr)
         );
-        panic!("command execution failed. command={:?}", cmd);
     }
+    return result;
 }
 
 fn cbindgen(rust_crate_dir: &str, c_output_path: &str, c_struct_names: Vec<String>) {
@@ -108,7 +150,7 @@ include = [{}]
     debug!("cbindgen config: {}", config);
 
     let mut config_file = tempfile::NamedTempFile::new().unwrap();
-    config_file.write_all(config.as_bytes()).unwrap();
+    std::io::Write::write_all(&mut config_file, config.as_bytes()).unwrap();
     debug!("cbindgen config_file: {:?}", config_file);
 
     let canonical = Path::new(rust_crate_dir)
@@ -121,21 +163,27 @@ include = [{}]
         path = &path[r"\\?\".len()..];
     }
 
-    execute_command(
-        &format!(
-            "cbindgen -v --config \"{}\" --output \"{}\"",
+    let res = execute_command(
+        "cbindgen",
+        &[
+            "-v",
+            "--config",
             config_file.path().to_str().unwrap(),
+            "--output",
             c_output_path,
-        ),
+        ],
         Some(path),
     );
+    if !res.status.success() {
+        panic!("cbindgen failed: {}", String::from_utf8_lossy(&res.stderr));
+    }
 }
 
 fn ffigen(
     c_path: &str,
     dart_path: &str,
     dart_class_name: &str,
-    llvm_path: &str,
+    llvm_path: &[String],
     llvm_compiler_opts: &str,
 ) {
     debug!(
@@ -159,12 +207,15 @@ fn ffigen(
         dart_path, dart_class_name, c_path, c_path,
     );
     if !llvm_path.is_empty() {
-        config = format!(
-            "{}
-        llvm-path:
-            - '{}'",
-            config, llvm_path
-        );
+        write!(
+            &mut config,
+            "
+        llvm-path:\n"
+        )
+        .unwrap();
+        for path in llvm_path {
+            write!(&mut config, "           - '{}'\n", path).unwrap();
+        }
     }
 
     if !llvm_compiler_opts.is_empty() {
@@ -179,22 +230,46 @@ fn ffigen(
     debug!("ffigen config: {}", config);
 
     let mut config_file = tempfile::NamedTempFile::new().unwrap();
-    config_file.write_all(config.as_bytes()).unwrap();
+    std::io::Write::write_all(&mut config_file, config.as_bytes()).unwrap();
     debug!("ffigen config_file: {:?}", config_file);
 
     // NOTE please install ffigen globally first: `dart pub global activate ffigen`
-    execute_command(
-        &format!(
-            "dart pub global run ffigen --config \"{}\"",
-            config_file.path().to_str().unwrap()
-        ),
+    let res = execute_command(
+        "dart",
+        &[
+            "pub",
+            "global",
+            "run",
+            "ffigen",
+            "--config",
+            config_file.path().to_str().unwrap(),
+        ],
         None,
     );
+    if !res.status.success() {
+        let err = String::from_utf8_lossy(&res.stderr);
+        if err.contains("Couldn't find dynamic library in default locations.") {
+            error!(
+                "
+ffigen could not find LLVM.
+Please supply --llvm-path to flutter_rust_bridge_codegen, e.g.:
+
+    flutter_rust_bridge_codegen .. --llvm-path <path_to_llvm>
+"
+            );
+            std::process::exit(Failures::FfigenLlvm as _);
+        }
+        panic!("ffigen failed:\n{}", err);
+    }
 }
 
 pub fn format_rust(path: &str) {
     debug!("execute format_rust path={}", path);
-    execute_command(&format!("rustfmt \"{}\"", path), None);
+    let res = execute_command("rustfmt", &[path], None);
+    if !res.status.success() {
+        error!("rustfmt failed: {}", String::from_utf8_lossy(&res.stderr));
+        std::process::exit(Failures::Rustfmt as _);
+    }
 }
 
 pub fn format_dart(path: &str, line_length: i32) {
@@ -202,12 +277,16 @@ pub fn format_dart(path: &str, line_length: i32) {
         "execute format_dart path={} line_length={}",
         path, line_length
     );
-    execute_command(
-        &format!(
-            "dart format \"{}\" --line-length {}",
-            path,
-            &line_length.to_string(),
-        ),
+    let res = execute_command(
+        "dart",
+        &["format", path, "--line-length", &line_length.to_string()],
         None,
     );
+    if !res.status.success() {
+        error!(
+            "dart format failed: {}",
+            String::from_utf8_lossy(&res.stderr)
+        );
+        std::process::exit(Failures::Dartfmt as _);
+    }
 }

--- a/frb_codegen/src/config.rs
+++ b/frb_codegen/src/config.rs
@@ -45,7 +45,7 @@ pub struct RawOpts {
     pub skip_add_mod_to_lib: bool,
     /// Path to the installed LLVM
     #[structopt(long)]
-    pub llvm_path: Option<String>,
+    pub llvm_path: Option<Vec<String>>,
     /// LLVM compiler opts
     #[structopt(long)]
     pub llvm_compiler_opts: Option<String>,
@@ -62,7 +62,7 @@ pub struct Opts {
     pub class_name: String,
     pub dart_format_line_length: i32,
     pub skip_add_mod_to_lib: bool,
-    pub llvm_path: String,
+    pub llvm_path: Vec<String>,
     pub llvm_compiler_opts: String,
     pub manifest_path: String,
 }
@@ -105,7 +105,20 @@ pub fn parse(raw: RawOpts) -> Opts {
         class_name,
         dart_format_line_length: raw.dart_format_line_length.unwrap_or(80),
         skip_add_mod_to_lib: raw.skip_add_mod_to_lib,
-        llvm_path: raw.llvm_path.unwrap_or_else(|| "".to_string()),
+        llvm_path: raw.llvm_path.unwrap_or_else(|| {
+            vec![
+                "/opt/homebrew/opt/llvm".to_owned(), // Homebrew root
+                "/usr/local/opt/llvm".to_owned(),    // Homebrew x86-64 root
+                // Possible Linux LLVM roots
+                "/usr/lib/llvm-9/lib".to_owned(),
+                "/usr/lib/llvm-10/lib".to_owned(),
+                "/usr/lib/llvm-11/lib".to_owned(),
+                "/usr/lib/llvm-12/lib".to_owned(),
+                "/usr/lib/llvm-13/lib".to_owned(),
+                "/usr/lib/".to_owned(),
+                "/usr/lib64/".to_owned(),
+            ]
+        }),
         llvm_compiler_opts: raw.llvm_compiler_opts.unwrap_or_else(|| "".to_string()),
         manifest_path,
     }

--- a/frb_codegen/src/main.rs
+++ b/frb_codegen/src/main.rs
@@ -7,6 +7,7 @@ use pathdiff::diff_paths;
 use structopt::StructOpt;
 
 use crate::api_types::ApiType;
+use crate::commands::ensure_tools_available;
 use crate::config::RawOpts;
 use crate::others::*;
 use crate::utils::*;
@@ -25,6 +26,8 @@ mod utils;
 
 fn main() {
     env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();
+
+    ensure_tools_available();
 
     let config = config::parse(RawOpts::from_args());
     info!("Picked config: {:?}", &config);
@@ -98,7 +101,7 @@ fn main() {
                 temp_dart_wire_file.path().as_os_str().to_str().unwrap(),
                 &config.dart_wire_class_name(),
                 c_struct_names,
-                &config.llvm_path,
+                &config.llvm_path[..],
                 &config.llvm_compiler_opts,
             );
         },

--- a/justfile
+++ b/justfile
@@ -48,4 +48,4 @@ clean:
     cd {{frb_flutter}} && flutter clean
     cd {{frb_flutter}}/rust && cargo clean
 
-# vim:expandtab:tabstop=4:shiftwidth=4
+# vim:expandtab:ts=4:sw=4

